test/Dialect/Comb/canonicalization.mlir

```diff
@@ -1533,3 +1533,41 @@ hw.module @OrMuxSameTrueValueAndZero(in %tag_0: i1, in %tag_1: i1, in %tag_2: i1
   %4 = comb.or bin %0, %1, %2, %3 : i4
   hw.output %4 : i4
 }
+
+// CHECK-LABEL: hw.module @combineOppositeCmpIntoConstant
+// CHECK-NEXT: %[[FALSE:.+]] = hw.constant false
+// CHECK-NEXT: %[[TRUE:.+]] = hw.constant true
+// CHECK-NEXT: hw.output %[[TRUE]], %[[FALSE]], %[[TRUE]], %[[FALSE]], %[[TRUE]], %[[FALSE]], %[[TRUE]], %[[FALSE]], %[[TRUE]], %[[FALSE]]
+hw.module @combineOppositeCmpIntoConstant(in %tag_0: i4, in %tag_1: i4, out o0: i1, out o1: i1, out o2: i1, out o3: i1, out o4: i1, out o5: i1, out o6: i1, out o7: i1, out o8: i1, out o9: i1) {
+  %op_ne = comb.icmp bin ne %tag_0, %tag_1 : i4
+  %op_eq = comb.icmp bin eq %tag_0, %tag_1 : i4
+  %eq_ne_or = comb.or %op_ne, %op_eq : i1
+  %eq_ne_and = comb.and %op_ne, %op_eq : i1
+
+  %op_slt = comb.icmp bin slt %tag_0, %tag_1 : i4
+  %op_sge = comb.icmp bin sge %tag_0, %tag_1 : i4
+  %slt_sge_or = comb.or %op_slt, %op_sge : i1
+  %slt_sge_and = comb.and %op_slt, %op_sge : i1
+
+  %op_sle = comb.icmp bin sle %tag_0, %tag_1 : i4
+  %op_sgt = comb.icmp bin sgt %tag_0, %tag_1 : i4
+  %sle_sgt_or = comb.or %op_sle, %op_sgt : i1
+  %sle_sgt_and = comb.and %op_sle, %op_sgt : i1
+
+  %op_ult = comb.icmp bin ult %tag_0, %tag_1 : i4
+  %op_uge = comb.icmp bin uge %tag_0, %tag_1 : i4
+  %ult_uge_or = comb.or %op_ult, %op_uge : i1
+  %ult_uge_and = comb.and %op_ult, %op_uge : i1
+
+  %op_ule = comb.icmp bin ule %tag_0, %tag_1 : i4
```

Comment:
The only-valid-under-bin property means you should have a negative test for non-bin icmp ops.

Reply:
Added